### ISSUE TITLE
libc should just be another library

### ISSUE THREAD

**Issue by Ericson2314** [NixOS/nixpkgs#26004]
Right now, libc is special: instead of just being another library dependency, it is passed to cc-wrapper where a bunch of magic happens for dynamic linking. I rather it just be another library with the caveat that if you don't depend on it dynamic linking won't work. Worst case, some of that dynamic-linking logic would go from the wrapper to the setup hook. `mkDerivation` would add that dependency by default, but packages (e.g. libc itself for obvious reasons) would opt out.

Relatedly, when bootstrapping (a new stdenv from a tarball or a cross stdenv), we current build "static" versions of the compiler that's aren't wrapped with libc, and then build libc, and then build the compiler again. At the very last, I'd like to just re-wrap the old compiler with libc and avoid the rebuild, but if we're not wrapping libc too, then let's just build *and* wrap the compiler once.

I assume there's a reason for all the double wrapping and double building we do now, but I don't know what it is.

CC @copumpkin @vcunat @DavidEGrayson @elitak @dmjio @eternaleye

**Comment by Ericson2314** [comment]
@DavidEGrayson thanks for that info! Sounds like to start, we'll still need to build GCC twice, but that shouldn't impede the cc-wrapper side of things--i.e. maybe the second GCC will build fine if it has libc as a normal dep instead of from the first gcc's wrapper script?

**Comment by eternaleye** [comment]
@DavidEGrayson 

> Since libstdc++ is a component of GCC and I am not sure if you can compile it on its own

It's worth noting here that it _is_ possible to split the build process of GCC considerably further than even that. In particular, Exherbo has managed to split out all of the following into distinct packages:

- libatomic
- libcilkrts
- libgcc (for libgcc_s.so)
- libgfortran
- libgomp
- libquadmath
- libstdc++

All of these can be found in [this directory](http://git.exherbo.org/arbor.git/tree/packages/sys-libs) of the `arbor` repository; `gcc` itself (with `libgcov` still bundled, as well as `crt*.o and `libgcc.a`) can be found [here](http://git.exherbo.org/arbor.git/tree/packages/sys-devel/gcc).

Exherbo _does_ still require rebuilding, however - the `threads` option on GCC requires the ability for the build process of the GCC package (in libgcc.a, I think) to link against the target libc, so one builds first without `threads`, then the target libc, then with `threads`. Building `libgcc.a` has not yet been factored out.

**Comment by Ericson2314** [comment]
Oooo that's so tantalizingly close.

**Comment by eternaleye** [comment]
Yup. At very least, it _does_ allow the second GCC to treat libc as a normal dep, right out of the gate.

**Comment by vcunat** [comment]
The (main) motivation is reducing the amount of compiling when bootstrapping?

**Comment by Ericson2314** [comment]
@vcunat Yes that, because when cross compiling there's no cache to rely on. But also simplicity for simplicity's sake---I'm convinced that with enough blood sweat and tears bootstrapping can become dramatically simpler.

**Comment by vcunat** [comment]
There are some cross tools in cache, I believe, mainly from http://hydra.nixos.org/jobset/nixpkgs/cross-trunk.  If there are more cross-scenarios that are "considered supported/maintained", I think it would be good to put a few basic packages in there for them.

It's not just for having binaries, but also for easier keeping an eye on breakages (and it simplifies finding when they started).

**Comment by Ericson2314** [comment]
Oh, I'm thinking of soon proposing we try to cross build the whole world for some platform, now that the infrastructure is getting good enough for that to be a realistic goal.

But for embedded stuff and other things, I still suspect people will be forced to fine-tune crossSystem breaking the cache.

**Comment by Ericson2314** [comment]
https://github.com/NixOS/nixpkgs/issues/38451 would be easier to write if libc and the C compiler's runtime libs were `propagatedBuildInputs`, which is implied by this.

**Comment by Ericson2314** [comment]
https://github.com/richfelker/musl-cross-make is some very interesting prior art.

**Comment by DavidEGrayson** [comment]
Yes, I referred to that project when making my own [cross-compilers for Linux](https://github.com/DavidEGrayson/nixcrpkgs/tree/master/linux), though I just stripped it down to the bare essentials and I was only concerned with static linking so I didn't try to get dynamic linking working.

**Comment by Ericson2314** [comment]
Still do

**Comment by Ericson2314** [comment]
Still important to me!

**Comment by HariAmoor-professional** [comment]
Can we do this with `zig cc`, which uses [Ziglang](https://zig.news/kristoff/compile-a-c-c-project-with-zig-368) to bootstrap CC and CXX? This adds a lot of benefits for hermeticity.

**Comment by HariAmoor-professional** [comment]
@Atemu The reason I figured it would is that a derivation could just download the binary for the underlying architecture instead of having to compile it

I could be wrong

**Comment by Ericson2314** [comment]
Zig was inspired by Nixpkgs, so I think we're good.

Clang already does a good job https://github.com/NixOS/nixpkgs/pull/132343 will help GCC. and @emilytrau's bootstrapping work will make everything easier too.

**Comment by Ericson2314** [comment]
This got done in 5c1955ff1404eddf3c0379a0573aabe83ac5bca3!
